be/src/formats/parquet/stored_column_reader.cpp

```diff
@@ -671,4 +698,282 @@ void StoredColumnReader::update_read_context(size_t records_read) {
     }
 }
 
+size_t StoredColumnReader::get_level_to_decode_batch_size(size_t row, size_t num_values_left_in_cur_page,
+                                                          size_t decoded, size_t parsed) {
+    constexpr size_t min_level_batch_size = 4096;
+    constexpr size_t max_level_batch_size = 1024 * 1024;
+    size_t levels_remaining = decoded - parsed;
+    if (row <= levels_remaining) {
+        return 0;
+    }
+
+    size_t levels_to_decode = std::max(min_level_batch_size, row - levels_remaining);
+    levels_to_decode = std::min(levels_to_decode, num_values_left_in_cur_page - levels_remaining);
+    levels_to_decode = std::min(levels_to_decode, max_level_batch_size);
+    return levels_to_decode;
+}
+
+size_t StoredColumnReader::count_not_null(level_t* def_levels, size_t num_parsed_levels, level_t max_def_level) {
+    size_t count = 0;
+    for (int i = 0; i < num_parsed_levels; ++i) {
+        level_t def_level = def_levels[i];
+        if (def_level == max_def_level) {
+            count++;
+        }
+    }
+    return count;
+}
+
+Status StoredColumnReader::read_range(const Range<uint64_t>& range, const Filter* filter,
+                                      ColumnContentType content_type, Column* dst) {
+    // reset() to prepare levels for new reading
+    reset();
+    if (_read_cursor < range.begin()) {
+        RETURN_IF_ERROR(_skip(range.begin() - _read_cursor));
+    }
+    return _read(range, filter, content_type, dst);
+}
+
+Status StoredColumnReader::_skip(uint64_t rows_to_skip) {
+    uint64_t skipped_row = 0;
+    while (skipped_row < rows_to_skip) {
+        uint64_t batch_to_skip = std::min(rows_to_skip - skipped_row, (uint64_t)10000);
```

Comment:
why do `min()` here?

Reply:
May be skip is very big, and we may need levels when skipping, we’d better skip batch then batch